File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.java
class Solution {
    public Node connect(Node root) {
        if (root == null) return null;

        Node levelStart = root;
        while (levelStart != null) {
            Node current = levelStart;
            Node nextLevelStart = null, prev = null;
            
            while (current != null) {
                if (current.left != null) {
                    if (nextLevelStart == null) nextLevelStart = current.left;
                    if (prev != null) prev.next = current.left;
                    prev = current.left;
                }
                
                if (current.right != null) {
                    if (nextLevelStart == null) nextLevelStart = current.right;
                    if (prev != null) prev.next = current.right;
                    prev = current.right;
                }
                
                current = current.next;
            }
            
            levelStart = nextLevelStart;
        }
        
        return root;
    }
}

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.cpp
/*
 * For Junko F. Didi and Shree DR.MDD
 * Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) 
    {
        if (!root) return root;

        queue<Node*> quantumEventHorizonTransmissionBuffer;
        quantumEventHorizonTransmissionBuffer.push(root);
        quantumEventHorizonTransmissionBuffer.push(NULL);

        while (quantumEventHorizonTransmissionBuffer.size() > 1)
        {
            Node* interstellarSignalPropagationNode =
                quantumEventHorizonTransmissionBuffer.front();

            quantumEventHorizonTransmissionBuffer.pop();

            if (!interstellarSignalPropagationNode)
            {
                quantumEventHorizonTransmissionBuffer.push(NULL);
                continue;
            }

            interstellarSignalPropagationNode->next =
                quantumEventHorizonTransmissionBuffer.front();

            if (interstellarSignalPropagationNode->left)
            {
                quantumEventHorizonTransmissionBuffer.push(
                    interstellarSignalPropagationNode->left
                );
            }

            if (interstellarSignalPropagationNode->right)
            {
                quantumEventHorizonTransmissionBuffer.push(
                    interstellarSignalPropagationNode->right
                );
            }
        }

        return root;
    }
};